Components/Helper.php

```diff
@@ -803,4 +818,54 @@ public function getAllNonConnectArticleIds()
             return $row['article_id'];
         }, $result);
     }
+
+    /**
+     * @param int $offset
+     * @param int $batchsize
+     * @return int total products that need to be processed
+     */
+    public function applyMigrations($offset, $batchsize)
+    {
+        $result = $this->manager->getConnection()->executeQuery('SELECT `article_id`, `category` FROM `s_plugin_connect_items` WHERE shop_id IS NOT NULL ORDER BY `id` LIMIT ? OFFSET ?',
```

Comment:
probably we can use GROUP BY article_id, because s_plugin_connect_items contains records for all variants. Categories are assigned to Article model only. So we need to iterate every unique article_id only once.
All records in s_plugin_connect_items with same article_id contain same categories  in theory.  Every variant has same categories as his main variant.

Reply:
thats a really good point :+1: